From d02d5742a0186bf5ee5685333de90961fa0b9bc3 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Thu, 17 Jun 2021 18:43:48 +0200
Subject: [PATCH] PONRTSYS-9036: Add function tcf_vlan_push_prio()

This adds the new function tcf_vlan_push_prio() and makes use of it.
On kernel 4.19 used by LGM this function is already integrated.
This allows to remove some of the ifdefs.
---
 .../net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c | 22 +++-------------------
 include/net/tc_act/tc_vlan.h                       |  5 +++++
 2 files changed, 8 insertions(+), 19 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
index 823363eabfc8..dc177e356641 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
@@ -266,13 +266,7 @@ static void dscp_parse(struct net_device *dev,
 					f->key);
 
 	dscp_value = (key->tos & IP_TOS_TO_DSCP) >> 2;
-#if (KERNEL_VERSION(4, 15, 0) > LINUX_VERSION_CODE)
-	act->dscp_pcp_map[dscp_value] =
-		((struct tcf_vlan *)a)->tcfv_push_prio;
-#else
-	act->dscp_pcp_map[dscp_value] =
-		((struct tcf_vlan *)a)->vlan_p->tcfv_push_prio;
-#endif
+	act->dscp_pcp_map[dscp_value] = tcf_vlan_push_prio(a);
 	act->dscp_pcp_map_valid[dscp_value] = true;
 	act->prio[act->push_n] = DERIVE_FROM_DSCP;
 	netdev_dbg(dev, "DSCP detected, value: %u, prio: %u\n",
@@ -284,11 +278,7 @@ static void vlan_action_modify_prio_parse(struct net_device *dev,
 					  enum tc_flower_vlan_tag tag,
 					  int *prio)
 {
-#if (KERNEL_VERSION(4, 15, 0) > LINUX_VERSION_CODE)
-	*prio = ((struct tcf_vlan *)a)->tcfv_push_prio;
-#else
-	*prio = ((struct tcf_vlan *)a)->vlan_p->tcfv_push_prio;
-#endif
+	*prio = tcf_vlan_push_prio(a);
 
 	if (*prio)
 		return;
@@ -372,13 +362,7 @@ static int vlan_action_parse(struct net_device *dev,
 		if (tcf_vlan_action(a) == TCA_VLAN_ACT_PUSH) {
 			act->act |= DP_VLAN_ACT_PUSH;
 
-#if (KERNEL_VERSION(4, 15, 0) > LINUX_VERSION_CODE)
-			act->prio[act->push_n] =
-				((struct tcf_vlan *)a)->tcfv_push_prio;
-#else
-			act->prio[act->push_n] =
-				((struct tcf_vlan *)a)->vlan_p->tcfv_push_prio;
-#endif
+			act->prio[act->push_n] = tcf_vlan_push_prio(a);
 			netdev_dbg(dev, "PUSH action detected, push_n: %d, act: %x\n",
 				   act->push_n + 1, act->act);
 		} else if (tcf_vlan_action(a) == TCA_VLAN_ACT_MODIFY) {
diff --git a/include/net/tc_act/tc_vlan.h b/include/net/tc_act/tc_vlan.h
index dd7a90c1ef46..98098c153ab2 100644
--- a/include/net/tc_act/tc_vlan.h
+++ b/include/net/tc_act/tc_vlan.h
@@ -54,6 +54,11 @@ static inline __be16 tcf_vlan_push_proto(const struct tc_action *a)
 	return to_vlan(a)->tcfv_push_proto;
 }
 
+static inline u8 tcf_vlan_push_prio(const struct tc_action *a)
+{
+	return to_vlan(a)->tcfv_push_prio;
+}
+
 static inline u8 tcf_vlan_push_flags(const struct tc_action *a)
 {
 	u8 tcfv_push_flags;
